ports/atmel-samd/tick.c

```diff
@@ -52,21 +52,34 @@ void SysTick_Handler(void) {
 
 void tick_init() {
     uint32_t ticks_per_ms = common_hal_mcu_processor_get_frequency() / 1000;
-    SysTick_Config(ticks_per_ms);
+    SysTick_Config(ticks_per_ms-1);
     NVIC_EnableIRQ(SysTick_IRQn);
 }
 
 void tick_delay(uint32_t us) {
     uint32_t ticks_per_us = common_hal_mcu_processor_get_frequency() / 1000 / 1000;
-    uint32_t us_between_ticks = SysTick->VAL / ticks_per_us;
-    uint64_t start_ms = ticks_ms;
+    uint32_t us_until_next_tick = SysTick->VAL / ticks_per_us;
+    uint32_t start_tick;
     while (us > 1000) {
-        while (ticks_ms == start_ms) {}
-        us -= us_between_ticks;
-        start_ms = ticks_ms;
-        us_between_ticks = 1000;
+        // check if interrupts are disabled
+        if(__get_PRIMASK())
+           return; // if not just return
+        start_tick=SysTick->VAL;  // wait for SysTick->VAL to  RESET
+        while (SysTick->VAL < start_tick) {}
+        us -= us_until_next_tick;
+        us_until_next_tick = 1000;
+    }
+    if(us&&(us < us_until_next_tick)){
+        while (SysTick->VAL > ((us_until_next_tick - us) * ticks_per_us)) {}
+    }
+    else {
+        start_tick=SysTick->VAL;  // wait for SysTick->VAL to  RESET
+        while (SysTick->VAL < start_tick) {}
+        us -= us_until_next_tick;
+        if(us){
+            while (SysTick->VAL > ((1000 - us) * ticks_per_us)) {}
```

Comment:
I think this line is all we need after the loop above with the changed condition. We'll exit the loop when we no longer need to track ticks.

I don't think it handles when the timer wraps around between exiting the loop and hitting this one though. In that case you'll end up with an extra millisecond I believe.

Reply:
I agree - so do we care if it adds a millisecond? If so, I need to figure out something more robust here.

Reply:
Nah, lets get this in and see if it comes up.